### PR TITLE
profile activity for code review interview question

### PR DESCRIPTION
**Pull request by wesley-alcock-hs**
Take a look at my code!! Let me know what you think. Thanks!

### CODE REVIEW

File: ProfileActivity.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    
+    @Override
+    public void onAttach(Activity activity) {
+        super.onAttach(activity);
+        mParentActivity = activity;
+    }
+    
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        socialProfileList.add(getIntent().getExtras().getParcelable("profileList"));
```
Comment (bluebery) [inline ProfileActivity.java]: Please check if "profileList" exists

File: ProfileActivity.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import butterknife.InjectView;
+import kotlin.Pair;
+
+public final class ProfileListFragment extends Fragment implements SomeInterface {
+    
+    @Inject
+    private ManagerImpl manager;
+    private TextView mCountText;  
+  
+    protected LinkedList<SocialProfile> socialProfileList = new LinkedList<SocialProfile>();
```
Comment (bluebery) [inline ProfileActivity.java]: Does this need to be protected?

File: ProfileActivity.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        super.onCreateView(inflater, container, savedInstanceState)
+        View view = inflater.inflate(R.id.profile_screen, container, false);
+        return view;
+    }
+    
+    public void onStart() {
+        mCountText = (TextView) findViewById(R.id.text_count);
+        RecyclerView recyclingView = (RecyclerView) findViewById(R.id.recycler_view);
+        
+        mCountText.setText("Count of Profiles " + socialProfiles.size());
```
Comment (bluebery) [inline ProfileActivity.java]: Please check if mCountText is null

File: ProfileActivity.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    }
+ 
+    public boolean onCreateOptionsMenu(Menu menu) {
+        super.onCreateOptionsMenu(menu)
+    }
+
+    @Override
+    public boolean onOptionsItemSelected(MenuItem item) {
+        switch (item.getItemId()) {
+            case R.id.menu_compare:
```
Comment (bluebery) [inline ProfileActivity.java]: Please add default:

File: ProfileActivity.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    }
+    
+    class SocialProfile implements Parcelable {
+        // Parcelable Implementation Omitted
+        // A List of Interests and their rank
+        // Example: [{1, "MUSIC"}, {2, "FOOD"}, ... ]
+        public List<Pair<Integer, Interests>> interestsRank;
+    }
+
+    private enum Interests {
```
Comment (bluebery) [inline ProfileActivity.java]: Please place at the start.

File: ProfileActivity.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+import java.util.LinkedList;
+import java.util.List;
+import javax.inject.Inject;
+import butterknife.InjectView;
+import kotlin.Pair;
+
+public final class ProfileListFragment extends Fragment implements SomeInterface {
+    
+    @Inject
+    private ManagerImpl manager;
```
Comment (bluebery) [inline ProfileActivity.java]: Dependency injected variables should be of default scope.

File: ProfileActivity.java (hunk 1 above)
Comment (wesley-alcock-hs) [inline ProfileActivity.java]: getArrayList

File: ProfileActivity.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
+        super.onCreateView(inflater, container, savedInstanceState)
+        View view = inflater.inflate(R.id.profile_screen, container, false);
+        return view;
+    }
+    
+    public void onStart() {
+        mCountText = (TextView) findViewById(R.id.text_count);
```
Comment (wesley-alcock-hs) [inline ProfileActivity.java]: butterknife in onviewcreated

File: ProfileActivity.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        RecyclerView recyclingView = (RecyclerView) findViewById(R.id.recycler_view);
+        
+        mCountText.setText("Count of Profiles " + socialProfiles.size());
+        recyclingView.setAdapter(new SocialProfilesAdapter(socialProfileList));
+        recyclingView.setLayoutManager(new LinearLayoutManager());
```
Comment (wesley-alcock-hs) [inline ProfileActivity.java]: might have to be before setadapter

File: ProfileActivity.java (hunk 3 above)
Comment (wesley-alcock-hs) [inline ProfileActivity.java]: use string xml

File: ProfileActivity.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    }
+        
+    /**
+     * Finds and returns the profiles in this fragment that is similar to the similarTo parameter.
+     * @param similarTo A collection of profiles to be similar to
+     */
+    public LinkedList<SocialProfile> similar(ArrayList<SocialProfile> similarTo) {
```
Comment (wesley-alcock-hs) [inline ProfileActivity.java]: return List<>

File: ProfileActivity.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        for (int i = 0; i < socialProfileList.size() - 1; i++) {
```
Comment (wesley-alcock-hs) [inline ProfileActivity.java]: use java comparison embedded in collections

File: ProfileActivity.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+                    socialProfileList.remove(i)
+                }
+            }
+        }
+        return socialProfileList;
+    }
+
+    public int compareInterests(SocialProfile s1, SocialProfile s2) {
+        int count = 0;
+        for (Pair<Integer, Interests> interestRank : s1.interestsRank) {
```
Comment (wesley-alcock-hs) [inline ProfileActivity.java]: use old school for loops instead of iterator

File: ProfileActivity.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+                    if (interestRank.getFirst() != interestRank.getFirst()) {
+                        count++;
+                    }
+                }
+            }
+        }
+        return count;
+    }
+    
+    class SocialProfile implements Parcelable {
```
Comment (wesley-alcock-hs) [inline ProfileActivity.java]: possibly private. could be static?

File: ProfileActivity.java (hunk 5 above)
Comment (wesley-alcock-hs) [inline ProfileActivity.java]: use intdef to use way less bytes

File: ProfileActivity.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+
+    private enum Interests {
+        MUSIC,
+        SPORTS,
+        SCIENCE,
+        MOVIES,
+        FOOD
+    }
+ 
+    public boolean onCreateOptionsMenu(Menu menu) {
```
Comment (wesley-alcock-hs) [inline ProfileActivity.java]: redundant

File: ProfileActivity.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        super.onCreateOptionsMenu(menu)
+    }
+
+    @Override
+    public boolean onOptionsItemSelected(MenuItem item) {
+        switch (item.getItemId()) {
```
Comment (wesley-alcock-hs) [inline ProfileActivity.java]: only use switch statements if more than 2 items. use if else

File: ProfileActivity.java (hunk 11 above)
Comment (wesley-alcock-hs) [inline ProfileActivity.java]: seems quite expensive

File: ProfileActivity.java [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+        return socialProfileList;
+    }
+
+    public int compareInterests(SocialProfile s1, SocialProfile s2) {
+        int count = 0;
+        for (Pair<Integer, Interests> interestRank : s1.interestsRank) {
+            for(Pair<Integer, Interests> interestRank2 : s2.interestsRank) {
+                if (interestRank2.getSecond() == interestRank.getSecond()) {
+                    if (interestRank.getFirst() != interestRank.getFirst()) {
+                        count++;
```
Comment (wesley-alcock-hs) [inline ProfileActivity.java]: break after count.
do something where for loops were in their own method. when got match return boolean, then calling method would do something

File: ProfileActivity.java (hunk 2 above)
Comment (wesley-alcock-hs) [inline ProfileActivity.java]: testinmg 123

File: ProfileActivity.java [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+    protected LinkedList<SocialProfile> socialProfileList = new LinkedList<SocialProfile>();
+    Activity mParentActivity;
+    
+    @Override
+    public void onAttach(Activity activity) {
+        super.onAttach(activity);
+        mParentActivity = activity;
+    }
+    
+    public void onCreate(Bundle savedInstanceState) {
```
Comment (neil-power-hs) [inline ProfileActivity.java]: some comment

File: ProfileActivity.java [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+import kotlin.Pair;
+
+public final class ProfileListFragment extends Fragment implements SomeInterface {
+    
+    @Inject
+    private ManagerImpl manager;
+    private TextView mCountText;  
+  
+    protected LinkedList<SocialProfile> socialProfileList = new LinkedList<SocialProfile>();
+    Activity mParentActivity;
```
Comment (liyucun) [inline ProfileActivity.java]: unused field

Comment (liyucun) [inline ProfileActivity.java]: you can always get parent activity by calling a method: getActivity()?

File: ProfileActivity.java [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+
+    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
+        super.onCreateView(inflater, container, savedInstanceState)
+        View view = inflater.inflate(R.id.profile_screen, container, false);
+        return view;
+    }
+    
+    public void onStart() {
+        mCountText = (TextView) findViewById(R.id.text_count);
+        RecyclerView recyclingView = (RecyclerView) findViewById(R.id.recycler_view);
```
Comment (liyucun) [inline ProfileActivity.java]: add another method for constructing recyclerView

File: ProfileActivity.java [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+import java.util.List;
+import javax.inject.Inject;
+import butterknife.InjectView;
+import kotlin.Pair;
+
+public final class ProfileListFragment extends Fragment implements SomeInterface {
+    
+    @Inject
+    private ManagerImpl manager;
+    private TextView mCountText;
```
Comment (liyucun) [inline ProfileActivity.java]: use butterknife

File: ProfileActivity.java (hunk 7 above)
Comment (liyucun) [inline ProfileActivity.java]: can be avoided by butterknife

File: ProfileActivity.java [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+        socialProfileList.add(getIntent().getExtras().getParcelable("profileList"));
+    }
+
+    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
+        super.onCreateView(inflater, container, savedInstanceState)
+        View view = inflater.inflate(R.id.profile_screen, container, false);
+        return view;
+    }
+    
+    public void onStart() {
```
Comment (liyucun) [inline ProfileActivity.java]: override

File: ProfileActivity.java (hunk 16 above)
Comment (liyucun) [inline ProfileActivity.java]: override

File: ProfileActivity.java [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+        super.onAttach(activity);
+        mParentActivity = activity;
+    }
+    
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        socialProfileList.add(getIntent().getExtras().getParcelable("profileList"));
+    }
+
+    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
```
Comment (liyucun) [inline ProfileActivity.java]: override

File: ProfileActivity.java (hunk 3 above)
Comment (liyucun) [inline ProfileActivity.java]: use format

File: ProfileActivity.java (hunk 9 above)
Comment (liyucun) [inline ProfileActivity.java]: 1. extract inner loop to another function to check if a social profile is similar to another one in a given list
2. don't remove a social profile while traversing it, use iterator

File: ProfileActivity.java (hunk 13 above)
Comment (liyucun) [inline ProfileActivity.java]: override

File: ProfileActivity.java [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+            }
+        }
+        return socialProfileList;
+    }
+
+    public int compareInterests(SocialProfile s1, SocialProfile s2) {
+        int count = 0;
+        for (Pair<Integer, Interests> interestRank : s1.interestsRank) {
+            for(Pair<Integer, Interests> interestRank2 : s2.interestsRank) {
+                if (interestRank2.getSecond() == interestRank.getSecond()) {
```
Comment (liyucun) [inline ProfileActivity.java]: can be extracted to another function

File: ProfileActivity.java (hunk 11 above)
Comment (liyucun) [inline ProfileActivity.java]: thinking about using functional style for this case

File: ProfileActivity.java (hunk 12 above)
Comment (liyucun) [inline ProfileActivity.java]: why not use another class?

File: ProfileActivity.java (hunk 6 above)
Comment (andy-fisher-hs) [inline ProfileActivity.java]: yhu

File: ProfileActivity.java (hunk 3 above)
Comment (andy-fisher-hs) [inline ProfileActivity.java]: list size is an integer so it's better to use string format %d

File: ProfileActivity.java [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+        View view = inflater.inflate(R.id.profile_screen, container, false);
+        return view;
+    }
+    
+    public void onStart() {
+        mCountText = (TextView) findViewById(R.id.text_count);
+        RecyclerView recyclingView = (RecyclerView) findViewById(R.id.recycler_view);
+        
+        mCountText.setText("Count of Profiles " + socialProfiles.size());
+        recyclingView.setAdapter(new SocialProfilesAdapter(socialProfileList));
```
Comment (andy-fisher-hs) [inline ProfileActivity.java]: Do we really want create new adapter instance each time at onStart? It's better to keep just one instance and update data then use dataSetChanged

File: ProfileActivity.java [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+            for (int j = 0; j < similarTo.size() - 1; j++) {
+                SocialProfile similarProfile = similarTo.get(j);
+                if (compareInterests(currentProfile, similarProfile) == 0) {
+                    socialProfileList.remove(i)
+                }
+            }
+        }
+        return socialProfileList;
+    }
+
```
Comment (andy-fisher-hs) [inline ProfileActivity.java]: There is too many level of loops to implement this function, if we want remove all social profile from socialProfileList with same conditions, it's better to take all interests from the array list simlarTo.
- Need more details about requirement of this function.

File: ProfileActivity.java [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+            }
+        }
+        return count;
+    }
+    
+    class SocialProfile implements Parcelable {
+        // Parcelable Implementation Omitted
+        // A List of Interests and their rank
+        // Example: [{1, "MUSIC"}, {2, "FOOD"}, ... ]
+        public List<Pair<Integer, Interests>> interestsRank;
```
Comment (andy-fisher-hs) [inline ProfileActivity.java]: Use HashMap or SparseArray instead List of Pair, if key is always integer, SparseArray is preferd

File: ProfileActivity.java (hunk 6 above)
Comment (snaderi) [inline ProfileActivity.java]: Inconsistent naming convention for member 'manager'.

File: ProfileActivity.java [hunk 26]
```diff
@@ -0,0 +1,103 @@
+package com.circle.profile.ProfileActivity;
+import com.hootsuite.manager.Manager;
+import com.circle.somelibrary.*;
+import java.util.LinkedList;
+import java.util.List;
+import javax.inject.Inject;
+import butterknife.InjectView;
+import kotlin.Pair;
+
+public final class ProfileListFragment extends Fragment implements SomeInterface {
```
Comment (snaderi) [inline ProfileActivity.java]: Did you mean to make a final class?

File: ProfileActivity.java (hunk 19 above)
Comment (snaderi) [inline ProfileActivity.java]: Text vs TextView?

File: ProfileActivity.java (hunk 2 above)
Comment (snaderi) [inline ProfileActivity.java]: Protected for final class?

File: ProfileActivity.java [hunk 27]
```diff
@@ -0,0 +1,103 @@
...
+    private ManagerImpl manager;
+    private TextView mCountText;  
+  
+    protected LinkedList<SocialProfile> socialProfileList = new LinkedList<SocialProfile>();
+    Activity mParentActivity;
+    
+    @Override
+    public void onAttach(Activity activity) {
+        super.onAttach(activity);
+        mParentActivity = activity;
```
Comment (snaderi) [inline ProfileActivity.java]: Check if implements SomeInterface.

File: ProfileActivity.java (hunk 1 above)
Comment (snaderi) [inline ProfileActivity.java]: Null checks.

File: ProfileActivity.java (hunk 7 above)
Comment (snaderi) [inline ProfileActivity.java]: Descriptive.

File: ProfileActivity.java (hunk 18 above)
Comment (snaderi) [inline ProfileActivity.java]: Descriptive.

File: ProfileActivity.java (hunk 3 above)
Comment (snaderi) [inline ProfileActivity.java]: Hardcoded strings.

File: ProfileActivity.java [hunk 28]
```diff
@@ -0,0 +1,103 @@
...
+        
+        mCountText.setText("Count of Profiles " + socialProfiles.size());
+        recyclingView.setAdapter(new SocialProfilesAdapter(socialProfileList));
+        recyclingView.setLayoutManager(new LinearLayoutManager());
+    }
+        
+    /**
+     * Finds and returns the profiles in this fragment that is similar to the similarTo parameter.
+     * @param similarTo A collection of profiles to be similar to
+     */
```
Comment (snaderi) [inline ProfileActivity.java]: More information needed.

File: ProfileActivity.java (hunk 5 above)
Comment (snaderi) [inline ProfileActivity.java]: Java enum?

File: ProfileActivity.java (hunk 13 above)
Comment (snaderi) [inline ProfileActivity.java]: not necessary

File: ProfileActivity.java [hunk 29]
```diff
@@ -0,0 +1,103 @@
...
+                }
+            }
+        }
+        return socialProfileList;
+    }
+
+    public int compareInterests(SocialProfile s1, SocialProfile s2) {
+        int count = 0;
+        for (Pair<Integer, Interests> interestRank : s1.interestsRank) {
+            for(Pair<Integer, Interests> interestRank2 : s2.interestsRank) {
```
Comment (snaderi) [inline ProfileActivity.java]: Naming of profiles awful.

File: ProfileActivity.java (hunk 12 above)
Comment (snaderi) [inline ProfileActivity.java]: Parcelable to be avoided. Pass by Id.

**Pull request closed without merging**